File: BinaryTree/python/allTreePaths.py
from typing import List, Optional

from node import Node


def all_tree_paths(root: Optional[Node]) -> List[List[int]]:
    if root is None:
        return []
    if root.left is None and root.right is None:
        return [[root.val]]

    paths: List[List[int]] = []
    for path in all_tree_paths(root.left):
        paths.append([root.val, *path])
    for path in all_tree_paths(root.right):
        paths.append([root.val, *path])
    return paths

File: BinaryTree/go/allTreePaths.go
package binarytree

// AllTreePaths returns every root-to-leaf path.
func AllTreePaths(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}

	var paths [][]int
	buildPaths(root, nil, &paths)
	return paths
}

func buildPaths(node *TreeNode, current []int, paths *[][]int) {
	if node == nil {
		return
	}

	current = append(current, node.Val)
	if node.Left == nil && node.Right == nil {
		path := make([]int, len(current))
		copy(path, current)
		*paths = append(*paths, path)
		return
	}

	buildPaths(node.Left, current, paths)
	buildPaths(node.Right, current, paths)
}
